module/zfs/zfs_vnops.c

```diff
@@ -1395,6 +1547,381 @@ zfs_clone_range(znode_t *inzp, uint64_t *inoffp, znode_t *outzp,
 	return (error);
 }
 
+
+/*
+ * This function is exactly the same as zfs_clone_range, except
+ * it is called with the range locks already held.
+ *
+ * It verifies that we hold locks for the src/dest clone range.
+ * It performs no unlocking of the locks.
+ *
+ * On success, the function return the number of bytes copied in *lenp.
+ * Note, it doesn't return how much bytes are left to be copied.
+ * On errors which are caused by any file system limitations or
+ * brt limitations `EINVAL` is returned. In the most cases a user
+ * requested bad parameters, it could be possible to clone the file but
+ * some parameters don't match the requirements.
+ */
+int
+zfs_clone_range_locked(znode_t *inzp, uint64_t *inoffp, znode_t *outzp,
+    uint64_t *outoffp, uint64_t *lenp, cred_t *cr, zfs_locked_range_t *src_zlr,
+    zfs_locked_range_t *dst_zlr)
+{
+	zfsvfs_t	*inzfsvfs, *outzfsvfs;
+	objset_t	*inos, *outos;
+	dmu_buf_impl_t	*db;
```

Comment:
I don't like to copy zfs_clone_range to a new function. For example: [encryption check](https://github.com/openzfs/zfs/blob/f9765b182eae1d031aa6559b22688329fff90fba/module/zfs/zfs_vnops.c#L1105) is not done here, because it was added to zfs_clone_range but not to this function. To copy all of the "new" checks to this function is bad and should be avoided if possible IMO. Based on your comment, I would suggest to pass a `boolen_t` to `zfs_clone_range`, which can be used to check if the lock is already held. Maybe it's also possible to check if the range locks are already held.